### PR TITLE
Optimized free memory handling

### PR DESCRIPTION
**Pull request by AnttiKauppila**
- Created a free list of unallocated blocks.
- Modified existing code to start using newly created list.

### CODE REVIEW

**Comment by AnttiKauppila** [comment]
@kjbracey-arm Could you review this?

**Comment by kjbracey-arm** [comment]
No actual commit message, saying what you've actually done?

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 1]
```diff
@@ -35,6 +36,26 @@ typedef enum mem_stat_update_t {
 
 static mem_stat_t *mem_stat_info_ptr = 0;
 
+typedef struct {
+    ns_list_link_t link;
+} hole_t;
+
+typedef NS_LIST_HEAD(hole_t, link) holes_list_t;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Don't think you need the typedef, so just NS_LIST_DEFINE would do. (That also initialises to empty, not that you need it).

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 2]
```diff
@@ -56,6 +77,7 @@ void ns_dyn_mem_init(uint8_t *heap, uint16_t h_size, void (*passed_fptr)(heap_fa
         heap += (sizeof(int) - temp_int);
         h_size -= (sizeof(int) - temp_int);
     }
+    memset(&holes_list, 0, sizeof(holes_list_t));
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: No no no. Zero-initting doesn't do anything useful, and you're initting it below anyway.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 3]
```diff
@@ -135,12 +161,7 @@ static int8_t ns_block_validate(int *block_start, int direction)
     int8_t ret_val = -1;
     int *end = block_start;
     int size_start = *end;
-    if (direction > 0) {
-        end += (1 + abs(size_start));
-    } else {
-        end -= (1 + abs(size_start));
-    }
-
+    end += (1 + abs(size_start));
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: So direction parameter no longer needed?

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 4]
```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
 static void *ns_dyn_mem_internal_alloc(const int16_t alloc_size, int direction)
 {
 #ifndef STANDARD_MALLOC
-    void *retval = 0;
+    int *retval = NULL;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: I think retval should stay void *. You're not manipulating it.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 5]
```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
...
     int data_size = convert_allocation_size(alloc_size);
+    int *hole_ptr = NULL;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Declare these where initialised.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 6]
```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
...
-                            if (direction > 0) {
-                                hole_ptr = ptr + data_size + 2;
-                            } else {
-                                hole_ptr = ptr - (data_size + 2);
-                            }
-                            int hole_size = block_data_size - data_size - 2;
-                            *hole_ptr = -(hole_size);
-                            if (direction > 0) {
-                                hole_ptr += (hole_size + 1);
+                int block_data_size = -(*ptr);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: The validation call is checking it's a valid block, either free or allocated. In this case we should be specifically checking it's free.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 7]
```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
...
+                    if (block_data_size >= (data_size + 2 + HOLE_T_SIZE)) {
+                        int hole_size = block_data_size - data_size - 2;
+
+                        //There is enough room for a new hole so create it first
+                        if ( direction > 0 ) {
+                            hole_ptr = ptr + data_size + 2;
+                        } else {
+                            hole_ptr = ptr;
+                        }
+                        if( direction > 0 ){
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Maybe worth a comment in the null else case, that we're keeping the existing descriptor, like in the other code.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 8]
```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
...
-                            ptr -= data_size;
-                            retval = ptr;
-                            ptr--;
+                        *hole_ptr = -(hole_size);
+                        hole_ptr += (hole_size + 1);
+                        *hole_ptr = -(hole_size);
+                        hole_ptr -= hole_size;
+                        if( direction < 0 ){
+                            ptr += hole_size + 2;
+                            hole_ptr = NULL; //no need to update
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Not needed.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 9]
```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
...
-                        *ptr = data_size;
-                        if (direction > 0) {
-                            ptr++;
-                            retval = ptr;
-                            ptr += data_size;
-                        } else {
-                            ptr -= data_size;
-                            retval = ptr;
-                            ptr--;
+                        *hole_ptr = -(hole_size);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Simplify to:

           hole_ptr[0] = -hole_size;
           hole_ptr[1 + hole_size] = -hole_size;
           if (direction < 0) {
                 ptr += 1 + hole_size + 1;  // I like this better than +2 cos it shows the actual layout.
           }

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 10]
```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
...
+                        ns_list_remove(&holes_list, hole_from_block_start(ptr));
                     }
-                }
-
-                moved += (block_data_size + 2);
-                if (direction > 0) {
-                    ptr += block_data_size + 2;
-                } else {
-                    ptr -= (block_data_size + 2);
+                    *ptr = data_size;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Again,

          ptr[0] = +data_size;
          ptr[1 + data_size] = +data_size;
          retval = ptr + 1;

Not a fan of this "tape-seeking" behaviour on a single block.

File: test/libService/unittest/makefile_defines.txt [hunk 11]
```diff
@@ -15,4 +15,4 @@ INCLUDE_DIRS =\
   $(CPPUTEST_HOME)/include\
 
 CPPUTESTFLAGS = -D__thumb2__ -w
-CPPUTEST_CFLAGS += -std=gnu99
\ No newline at end of file
+CPPUTEST_CFLAGS += -std=gnu99
```
Comment (kjbracey-arm) [inline test/libService/unittest/makefile_defines.txt]: Can see what the change is here (thanks GitHub). Double-check

File: test/libService/unittest/nsdynmem/dynmemtest.cpp [hunk 12]
```diff
@@ -256,9 +282,9 @@ TEST(dynmem, diff_sizes)
     int i;
     // Should leave headroom for 2 pointers
     for (i=1; i<(size-8); i++) {
-        p = ns_dyn_mem_temporary_alloc(i);
+        p = ns_dyn_mem_temporary_alloc(i); //1 block, 1, 1, 1, 2, 2, 2, 2
         CHECK(p);
-        ns_dyn_mem_free(p);
+        ns_dyn_mem_free(p); //1 block hangs, 1 merge,
```
Comment (kjbracey-arm) [inline test/libService/unittest/nsdynmem/dynmemtest.cpp]: Stray comment

**Comment by AnttiKauppila** [comment]
Smoketests passed

**Pull request merged**